src/Hooks/Hook.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace Carnival\Hooks;
+
+use Carnival\Hooks\Event;
+use Illuminate\Support\Collection;
+
+class Hook {
+
+    protected Collection $hooks;
+
+    public function __construct() {
+        $this->hooks = collect();
+    }
+
+    function add (string $eventClass, callable $callback) : void {
+        data_set($this->hooks, $eventClass, data_get($this->hooks, $eventClass, collect())->push($callback));
+    }
+    
+    function remove (string $eventClass) : void {    
+        $this->hooks->forget($eventClass);
+    }
+    
+    function execute (Event $event) : void {
+        data_get($this->hooks, get_class($event), collect())->each(fn ($callback) => $callback($event));
+    }
+
+    /**
+     * Functions for testing
+     */
+    function count (string $eventClass) : int {
```

Comment:
This isn't really needed for testing actually. We can more strategically test that a hook has been registered or not by triggering its event and seeing if the side effect occurs.

Reply:
Alright, maybe lets work together on reworking these. The count system seemed explicit and straightforward.

Reply:
> The count system seemed explicit and straightforward.

With the drawback of including unit-test-specific code in the source, which is undesirable. Any callback with a side effect would be sufficient for this case. For example, the string one you already wrote would work. Or the simple number incrementing one I suggested (for testing the multi-hook case) would work for this as well, using only one of the two hooks.

For the unregistered case, call the hook and see that nothing happens.
For the one registered case, call the hook and see that its effect happens.
For the registered and then unregistered case, call the hook and see that its effect doesn't happen.
So on